### DISCUSSION TITLE
Touch support

### DISCUSSION THREAD

**Discussion by gaunap** [webdriverio/webdriverio#7997] [Technical Discussion]
If I understand the documentation correctly no touch actions are currently supported bei webdriver: https://webdriver.io/docs/api/browser/touchAction/

Is that right?
Is there a plan to implement this?
Or any workarounds?

Mobile testing is very important to us and not having the possibility to create touch actions in the browser means that we can't test many crucial scenarious.

**Comment by christian-bromann** [comment]
What makes you think this is not implemented for mobile?

**Comment by gaunap** [comment]
Have you checked the link? It states: _It is currently only available to native apps and can not be used to interact with webapps._

**Reply by christian-bromann** [reply]
Yes, this clearly tells me that it works for testing mobile native apps.

**Reply by gaunap** [reply]
Alas, we have to test websites. native application are not in our focus.
Will browser based applications be supported in future?

**Reply by christian-bromann** [reply]
Ok, now I understand you. You want to do web tests in the browser on a mobile device. I was a bit confused based on the first sentence you wrote:

> Mobile testing is very important to us

For web testing in general you will use the [`performactions`](https://webdriver.io/docs/api/webdriver#performactions) command which has things like `touch` or `pen` as input source.

**Reply by gaunap** [reply]
Thanks for the hint! I will try this out.
In the same sentence I also mentioned also 
> touch actions in the browser
But anyways, the _performactions_ approach looks like a workaround.

This is counter intuitive for a naive user. Why is there a _browser.touchAction_ API which doesn't work in the browser?
At least a hint/link at the documentation of wdio (touchAction) to use performactions (with payload {...}) instead would be very nice.
And my questions is still valid: will _browser.touchAction_  be supported in browsers in future?

**Reply by christian-bromann** [reply]
> will browser.touchAction be supported in browsers in future?

No, because that specific touchAction command is only support by Appium and not part of the WebDriver spec.

> This is counter intuitive for a naive user.

I agree. We have an open issue that suggests to create a common command which would for browser and mobile devices and use different commands in the background. In fact we already have a development in this space in #7214 which targets to build a common interface that can be used across environments.

**Comment by aleks-temp** [comment]
Hi @christian-bromann. 
On our web project we have a specific area of the screen, which zooms in and out separately from the whole screen. This happens 2 ways: 1. By Ctrl+Mouse scroll or 2. From the touch pad of the laptop - by touching and moving fingers in opposite directions.

I am trying to simulate the second case with the following code:

`await browser.performActions([
    {
      type: "pointer",
      id: "finger1",
      parameters: { "pointerType": "touch" },
      actions: [
        { type: "pointerMove", duration: 0, x: startX, y: startY },
        { type: "pointerDown", button: 0 },
        { type: "pause", duration: 500 },
        { type: "pointerMove", duration: 1000, origin: "pointer", x: -100, y: 0 },
        { type: "pointerUp", "button": 0 }
      ]
    },
    {
      type: "pointer",
      id: "finger2",
      parameters: { "pointerType": "touch" },
      actions: [
        { type: "pointerMove", duration: 0, x: startX, y: startY },
        { type: "pointerDown", button: 0 },
        { type: "pause", duration: 500 },
        { type: "pointerMove", duration: 1000, origin: "pointer", x: 100, y: 0 },
        { type: "pointerUp", "button": 0 }
      ]
    }
  ]);`

but the code zooms the whole screen, not just the specific area (which is on the pointed coordinates);

**Reply by aleks-temp** [reply]
Here is some more detailed code to demonstrate the issue:

`await browser.url('https://www.booking.com/searchresults.en-gb.html?label=en-bg-booking-desktop-MtC3OoaoU_ML6gxkUZ2GoAS652796015943%3Apl%3Ata%3Ap1%3Ap2%3Aac%3Aap%3Aneg%3Afi%3Atikwd-65526620%3Alp1001448%3Ali%3Adec%3Adm&gclid=CjwKCAjw9p24BhB_EiwA8ID5BpYIfe-UdRSTactu4nGj4nIZ85-yr_Pb_npRQFukEtS8INtKRNdXHRoCCDAQAvD_BwE&aid=2311236&dest_id=-838489&dest_type=city&group_adults=2&req_adults=2&no_rooms=1&group_children=0&req_children=0#map_opened');
    let map = await $('div.gm-style');
    let x = await map.getLocation('x');
    let y = await map.getLocation('y');
    let width = await map.getSize('width');
    let height = await map.getSize('height');
    let startX = parseInt(x + width / 2);
    let startY = parseInt(y + height / 2);
    await browser.performActions([
        {
            type: "pointer",
            id: "finger1",
            parameters: { "pointerType": "touch" },
            actions: [
                { type: "pointerMove", duration: 0, x: startX, y: startY },
                { type: "pointerDown", button: 0 },
                { type: "pause", duration: 500 },
                { type: "pointerMove", duration: 1000, origin: "pointer", x: -100, y: 0 },
                { type: "pointerUp", "button": 0 }
            ]
        },
        {
            type: "pointer",
            id: "finger2",
            parameters: { "pointerType": "touch" },
            actions: [
                { type: "pointerMove", duration: 0, x: startX, y: startY },
                { type: "pointerDown", button: 0 },
                { type: "pause", duration: 500 },
                { type: "pointerMove", duration: 1000, origin: "pointer", x: 100, y: 0 },
                { type: "pointerUp", "button": 0 }
            ]
        }
    ]);`

Gesture performed on the map have to zoom the map only. But it zooms the whole page
